src/liblink.c

```diff
@@ -0,0 +1,244 @@
+// Copyright 2018 (c) Xerpa
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "liblink.h"
+#include <pthread.h>
+
+struct _liblink_sock_t {
+  volatile enum liblink_state state;
+  pthread_t ioloop;
+  pthread_mutex_t mutex;
+  zloop_t *reactor;
+  zsock_t *sock;
+  zsock_t *push;
+  zsock_t *pull;
+  liblink_recv_fn recvfn;
+  void *recvfn_args;
+};
+
+static int _liblink_ioloop_recv (zloop_t *zloop, zsock_t *sock, void *args);
+
+static
+void _liblink_state_w (liblink_sock_t *socket)
+{
+  if (0 == pthread_mutex_lock(&socket->mutex))
+  {
+    if (socket->state == LIBLINK_STATE_RUNNING)
+    {
+      socket->state = LIBLINK_STATE_WAITING;
+      zloop_reader_end(socket->reactor, socket->sock);
+    }
+
+    pthread_mutex_unlock(&socket->mutex);
+  }
+}
+
+static
+void _liblink_state_r (liblink_sock_t *socket)
+{
+  if (0 == pthread_mutex_lock(&socket->mutex))
+  {
+    if (socket->state == LIBLINK_STATE_WAITING)
+    {
+      socket->state = LIBLINK_STATE_RUNNING;
+      zloop_reader(socket->reactor, socket->sock, _liblink_ioloop_recv, socket);
+    }
+
+    pthread_mutex_unlock(&socket->mutex);
+  }
+}
+
+static
+void _liblink_state_t (liblink_sock_t *socket)
+{
+  if (0 == pthread_mutex_lock(&socket->mutex))
+  {
+    if (socket->state == LIBLINK_STATE_RUNNING || socket->state == LIBLINK_STATE_WAITING)
+    {
+      socket->state = LIBLINK_STATE_QUITTING;
+      zloop_reader_end(socket->reactor, socket->pull);
+      zloop_reader_end(socket->reactor, socket->sock);
+    }
+
+    pthread_mutex_unlock(&socket->mutex);
+  }
+}
+
+static
+int _liblink_ioloop_recv (zloop_t *zloop, zsock_t *sock, void *args)
+{
+  liblink_sock_t *socket = (liblink_sock_t *) args;
+  zmsg_t *msg = zmsg_recv(sock);
+  LIBLINK_UNUSED(zloop);
+
+  if (msg != NULL)
+  {
+    if (0 != socket->recvfn(msg, socket->recvfn_args))
+    { _liblink_state_w(socket); }
```

Comment:
É isto que faz o _socket_ parar de encaminhar mensagens pro PID quando o `socket->recvfn` devolve `-1`, e este devolve `-1` quando o `enif_send` tenta mandar mensagens para um processo morto?

Reply:
neste PR, se isso acontecer, as msgs são todas descartadas. no futuro [não lembro exatamente onde], em caso de erro escrevendo no pid ele para de consumir e pode-se trocar o pid q recebe as mensagens.

Reply:
Mas, se eu chamar  `_liblink_state_r` depois que algumas mensagens foram enviadas para este _socket_ que estava pausado, ele não vai pegar as mensagens da fila interna e enviar para o PID, como no `nif_test.exs:97`?

Reply:
sim, vc vê algum problema? 

se o socket for `dealer`, e mecanismo de timeout provavelmente vai descartar aquela mensagem. se for um `router`, o request provavelmente vai ser processado mas não vai haver ninguém esperando e o router vai descartar a mensagem.